### PR TITLE
Issue #2042: browser-icons: Add disk cache for icons.

### PR DESCRIPTION
**Pull request by pocmo**
This caches the list of resources and the image on disk. There are two improvements over Fennec:
* We store all icon resource URLs now
* We store icons as WEBP which are much smaller than PNGs

This introduces a new third party dependency: "Jake Wharton's" DiskLruCache. It's a standalone version of the DiskLruCache of AOSP. I would like to not have this dependency but it makes no sense to re-implement that and this library is used A LOT: 
https://www.appbrain.com/stats/libraries/details/disklrucache/disk-lru-cache

We could think of moving this dependency into our tree - it's not big. That would at least prevent us from depending on pulling it from jcenter.

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### CODE REVIEW

File: components/browser/icons/src/main/java/mozilla/components/browser/icons/extension/IconMessage.kt [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    "apple-touch-icon image_src" to IconRequest.Resource.Type.APPLE_TOUCH_ICON,
+    "apple-touch-icon-precomposed" to IconRequest.Resource.Type.APPLE_TOUCH_ICON,
+    "og:image" to IconRequest.Resource.Type.OPENGRAPH,
+    "og:image:url" to IconRequest.Resource.Type.OPENGRAPH,
+    "og:image:secure_url" to IconRequest.Resource.Type.OPENGRAPH,
+    "twitter:image" to IconRequest.Resource.Type.TWITTER,
+    "msapplication-TileImage" to IconRequest.Resource.Type.MICROSOFT_TILE
+)
+
+private fun Map<String, IconRequest.Resource.Type>.reverseLookup(type: IconRequest.Resource.Type): String {
```
Comment (Amejia481) [inline components/browser/icons/src/main/java/mozilla/components/browser/icons/extension/IconMessage.kt]: Should we return a list here. I noticed on `typeMap` there are multiple keys for one given value, for instance:

`IconRequest.Resource.Type.OPENGRAPH` has ` ["og:image", "og:image:url", "og:image:secure_url"]`

`IconRequest.Resource.Type.APPLE_TOUCH_ICON` has `["apple-touch-icon","apple-touch-icon precomposed","apple-touch-icon image_src"]`

Comment (pocmo) [reply components/browser/icons/src/main/java/mozilla/components/browser/icons/extension/IconMessage.kt]: The actual value used is not important in this case as long as when parsing it's deserialized to `IconRequest.Resource.Type.OPENGRAPH` again.

**Comment by pocmo** [comment]
Updated the failing test. Now it should pass.

File: components/browser/icons/src/main/java/mozilla/components/browser/icons/extension/IconMessage.kt (hunk 1 above)
Comment (Amejia481) [reply components/browser/icons/src/main/java/mozilla/components/browser/icons/extension/IconMessage.kt]: 👍

**Pull request merged**